lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -174,7 +168,10 @@ def relationship_value_for(serializer, options = {})
         end
 
         def relationships_for(serializer)
-          serializer.associations.each_with_object({}) do |association, hash|
+          resource_type = resource_identifier_type_for(serializer)
+          requested_associations = fieldset.fields_for(resource_type) || '*'
+          include_tree = IncludeTree.from_include_args(requested_associations)
+          serializer.associations(include_tree).each_with_object({}) do |association, hash|
```

Comment:
Exactly. The `IncludeTree` is probably not a perfect name. It is basically a structure that describes a labelled tree in such a way that you can query it using wildcards.